README.md
# SubLoader

> ### About SubLoader
SubLoader allows you to load subtitles for your content automatically.

Special thanks to [OpenSubtitles.org](http://www.opensubtitles.org ) for providing a unique UA for Subloader.

![OpenSubtitles.org](http://static.opensubtitles.org/gfx/logo-transparent.png)

> ### Issues
The addon is currently in Alpha stage. There are some bugs and several features to be added.

If you use this addon and have installed the addon "service.autosubs" please disable it; otherwise, the subloader won't function correctly.

The subloader needs a valid IMDBNumber, and the filename must include the media quality to work correctly. If your content doesn't provide a valid IMDBNumber, the addon tries to do a query search. This search typically takes at least 10 seconds to complete.

Most likely your episodes won't have an IMDBNumber because Kodi scrapper does not provide it at an episode level.

For now, the addon is only compatible with Kodi 18.0 or higher.

> ### Development
SubLoader is a fork from the well-known autosubs addon. It also contains some code from a Jen base addon. All credits go to the original developers.

I'm not a programmer; all I do is trial and error to find a possible solution. So please if someone wants to help out and to push this addon forward, be my guest.


changelog.txt
0.1.6 - By Denim (05.05.2021)
- Bug fix on open subtitle dialog.

0.1.5 - By Denim (21.02.2021)
- Subloader now runs on KODI v19 (Special thanks to peno/64).
- Added python 3 compatibility.
- Some grammatical corrections (Special thanks to AlexNaga).

0.1.4 - By Denim (23.07.2019)
- Added Dutch language (Special thanks to Tinuz97).
- Clean up project files.
 
 0.1.3 - By Denim (21.07.2019)
- Added OS user login.
- Fixed debug option.
- Added more debug information.
- Massive code rewrite.
- Several optimizations.
- Added audio language exclusion.
- Added video addon exclusion (multi selection).
- Improved sorting settings.
- Improved settings layout.
- Added "Do nothing" option.
- Fixed typo (Special thanks to AlexNaga).

0.1.2 - By Denim (17.06.2019)
- Added a unique User Agent from Opensubtitles.org.

0.1.1 - By Denim (12.06.2019)
- Fixed release detection bug.
- Fixed typo.

0.1.0 - By Denim (10.06.2019)
- Fixed instalation bug.
- Fixed notifications bug.
- Started Beta convergence.
- Added default behavior selection.
- Added auto donwload any subtitle if no match found.
- Updated settings xml.
- Updated strings po.
- Release type detection improved.

0.0.9 - By Denim
- Release type detection improved.
- More clean up.

0.0.8 - By Denim
- Added query search if no valid IMDBNumber was found.
- Clean up.

0.0.7 - By Denim
- Added Portuguese language.
- Added ability to choose (subtitle search dialog) as a fallback.

0.0.6 - By Denim
- Added notifications if enabled.

0.0.5 - By Denim
- Massive improve on video release types detection.
- Added ability to choose strict subtitle matching.

0.0.4 - By Denim
- Several fixes.
- Improved release detection.
- Improved seasons and episodes matches.
- Removed translations (several changes were made).

0.0.3 - By Denim
- Bug fixes.
- Added 3º language option.
- Improved IMDB number detection.

0.0.2 - By Denim
- Added a custom delay to start the service.
- Changed player trigger. 
- The script is now only compatible with kodi 18 or higher.
- Added ability to actually donwload subtitles (Have used some code from an jen base addon all credits to the developer).

0.0.1 - By Denim
- Initial release.
- Forked service.autosubs.
- All credits to the original developers.


monitor.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

import os, sys, xbmc, time
from resources.lib.loadsub import loadsub
from resources.lib.exclusions import globalexclusion
from resources.lib.OSserver import OSusersetting, OSuser
from resources.lib.utils import setting, boolsetting, setsetting, localize, debug, debugsetting


#Main Monitor********************************************************************************************************************************************

class SubLoaderMonitor(xbmc.Monitor):


	def __init__(self):
		xbmc.Monitor.__init__(self)
		debug("Initalized main monitor")
		self.run = True

	def onSettingsChanged(self):

		if self.run:	

			if debugsetting():
				self.run = False
				debug('Debug settings changed')
				setsetting('debugcheck', value=setting('debug'))
				global closedebug
				closedebug = boolsetting('debug')
				debug('Debug initalized')
				if not closedebug:
					debug('Debug stopped', force = True)

			if OSusersetting():
				self.run = False
				debug('OS User settings changed')
				if OSuser():
					xbmc.executebuiltin('Notification("SubLoader", "%s", "%s",)' % (localize(32033), 4000))
					setsetting('OSusercheck', value=setting('OSuser'))
					setsetting('OSpasswordcheck', value=setting('OSpassword'))
					debug('OS Login: successful')
				else:
					xbmc.executebuiltin('Notification("SubLoader", "%s", "%s",)' % (localize(32034), 4000))
					setsetting('OSusercheck', value='nouser')
					setsetting('OSpasswordcheck', value='nopassword')
					debug('OS Login: unsuccessful')

		else:
			xbmc.sleep(1000)
			self.run = True


monitor = SubLoaderMonitor()


#Player monitor******************************************************************************************************************************************

class SubLoaderPlayer(xbmc.Player):

	def __init__(self, *args, **kwargs):
		xbmc.Player.__init__(self)
		debug('Initalized player monitor')
		self.run = True

	def onPlayBackStopped(self):
		debug('Playback stopped')
		self.run = True

	def onPlayBackEnded(self):
		debug('Playback ended')
		self.run = True

	def onAVStarted(self):
		self.run = True
		delay = int(setting('delay'))*1000

		if self.run:
			xbmc.sleep(delay)		
			if xbmc.Player().isPlayingVideo() and globalexclusion():
				self.run = False	
				if setting('default') == '0':
					debug('Default: automatic subtitles')
					loadsub()
				elif setting('default') == '1':
					debug('Default: opening search dialog')
					xbmc.executebuiltin('ActivateWindow(SubtitleSearch)')
				else:
					debug('Default: do nothing...')
			else:
				self.run = False


player = SubLoaderPlayer()


#Abort request*******************************************************************************************************************************************

closedebug = boolsetting('debug')

while not xbmc.Monitor().abortRequested():
	
	if xbmc.Monitor().waitForAbort(10):
		if closedebug:
			debug('Shutdown requested', force = True)
			debug('Player monitor stopped', force = True)
			debug('Main monitor stopped', force = True)
		del player
		del monitor
		break

#********************************************************************************************************************************************************


resources/lib/utils.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

import os, sys, xbmc, xbmcaddon, xbmcgui
import base64, gzip, xbmcvfs
import re, hashlib, json
import codecs, urllib, xbmcplugin, time

try:
    import StringIO
    loglevel = xbmc.LOGNOTICE
    py2 = True
except:
    import io
    loglevel = xbmc.LOGINFO
    py2 = False

langdict = {
        'Afrikaans': 'afr', 'Albanian': 'alb', 'Arabic': 'ara', 'Armenian': 'arm', 'Basque': 'baq',
        'Bengali': 'ben', 'Bosnian': 'bos', 'Breton': 'bre', 'Bulgarian': 'bul', 'Burmese': 'bur',
        'Catalan': 'cat', 'Chinese': 'chi', 'Croatian': 'hrv', 'Czech': 'cze', 'Danish': 'dan', 'Dutch': 'dut',
        'English': 'eng', 'Esperanto': 'epo', 'Estonian': 'est', 'Finnish': 'fin', 'French': 'fre',
        'Galician': 'glg', 'Georgian': 'geo', 'German': 'ger', 'Greek': 'ell', 'Hebrew': 'heb', 'Hindi': 'hin',
        'Hungarian': 'hun', 'Icelandic': 'ice', 'Indonesian': 'ind', 'Italian': 'ita', 'Japanese': 'jpn',
        'Kazakh': 'kaz', 'Khmer': 'khm', 'Korean': 'kor', 'Latvian': 'lav', 'Lithuanian': 'lit',
        'Luxembourgish': 'ltz', 'Macedonian': 'mac', 'Malay': 'may', 'Malayalam': 'mal', 'Manipuri': 'mni',
        'Mongolian': 'mon', 'Montenegrin': 'mne', 'Norwegian': 'nor', 'Occitan': 'oci', 'Persian': 'per',
        'Polish': 'pol', 'Portuguese': 'por', 'Portuguese(Brazil)': 'pob', 'Romanian': 'rum',
        'Russian': 'rus', 'Serbian': 'scc', 'Sinhalese': 'sin', 'Slovak': 'slo', 'Slovenian': 'slv',
        'Spanish': 'spa', 'Swahili': 'swa', 'Swedish': 'swe', 'Syriac': 'syr', 'Tagalog': 'tgl', 'Tamil': 'tam',
        'Telugu': 'tel', 'Thai': 'tha', 'Turkish': 'tur', 'Ukrainian': 'ukr', 'Urdu': 'urd'}




def addon():
        addon = xbmcaddon.Addon().getAddonInfo('id')
        return xbmcaddon.Addon(addon)


def name():
        return xbmcaddon.Addon().getAddonInfo('name')


def version():
        return xbmcaddon.Addon().getAddonInfo('version')


def localize(id):
        if py2:
            return addon().getLocalizedString(id).encode('utf-8')
        else:
            return addon().getLocalizedString(id)


def setting(setting):
        return addon().getSetting(setting)


def boolsetting(setting):
        return addon().getSetting(setting).lower() == "true"


def setsetting(setting, value):
        return addon().setSetting(setting, value)


def setboolsetting(setting, value):
        return addon().setSettingBool(setting, value)


def debug(msg, force = False):

        if force or boolsetting('debug'):
                try:
                        xbmc.log("#####[SubLoader]##### " + msg, loglevel)
                except UnicodeEncodeError:
                        xbmc.log("#####[SubLoader]##### " + msg, loglevel).encode( "utf-8", "ignore" )

debug('Loading %s version %s' % (name(), version()))


def videopath():
        return xbmc.Player().getPlayingFile()


def videosource():
        return xbmc.getInfoLabel('Player.Folderpath')


def fullvideosource():
        return xbmc.getInfoLabel('Player.Filenameandpath')


def debugsetting():

        if boolsetting('debug') != boolsetting('debugcheck'):
                return True
        return False


resources/lib/OSserver.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

import os, sys, xbmc
from resources.lib.utils import name, version, setting, boolsetting

try:
    import xmlrpc.client as xc
except:
    import xmlrpclib
    xc = xmlrpclib



OpenSubtitles = xc.Server('http://api.opensubtitles.org/xml-rpc', verbose=0)




#OpenSubtilte.org functions*******************************************************************************************************************************

def OSserver():
        user = ""
        password = ""
        ua = "kodi_%s_v%s" % (name().lower(), version())
        if boolsetting('OScustomuser'):
                if setting ('OSuser') != '' or setting('OSpassword') != '':
                        user = setting('OSuser')
                        password = setting('OSpassword')
                else:
                        user = "error"
                        password = "error"
        return OpenSubtitles.LogIn(user, password, 'en', ua)


def OSinfo():
        return OpenSubtitles.ServerInfo()['website_url']


def OStoken():
        return OSserver()['token']


def OSuser():
        return OSserver()['status'] == "200 OK"


def OSusersetting():
        if boolsetting('OScustomuser') and (setting('OSuser') != setting('OSusercheck') or setting('OSpassword') != setting('OSpasswordcheck')):
                return True
        return False

#*********************************************************************************************************************************************************


resources/lib/exclusions.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

import xbmc, datetime
from resources.lib import utils
from resources.lib.utils import setting, videosource, debug, videopath, boolsetting


# Exclusions*********************************************************************************************************************************************

def addonexclusion():

	if setting('excludeaddon'):
		exaddon = setting('excludeaddon').split(',')
		if all(videosource().find (v) <= -1 for v in exaddon):
			return True
		else:
			debug('Excluded: the source addon is excluded %s' % exaddon)
			return False
	else:
		return True


def wordsexclusion():

	if setting('excludewords'):
		words = setting('excludewords').split(',')
		if all(videopath().find (v) <= -1 for v in words):
			return True
		else:
			debug('Excluded: the video path have excluded words %s' % words)
			return False
	else:
		return True


def timeexclusion():

	time = xbmc.Player().getTotalTime()
	debug('Total time: %s' % str(datetime.timedelta(seconds=round(time))))
	timeexcluded = int(setting('excludetime'))*60
	if time > timeexcluded:
		return True
	else:
		debug('Excluded: the content time is lower than the exclusion')
		return False


def audioexclusion():

	if boolsetting('excludeaudio'):
		langs = []
		langs.append(utils.langdict[setting('excludeaudiolang1')])
		if not setting('excludeaudiolang2') == "-----":
			langs.append(utils.langdict[setting('excludeaudiolang2')])
		if not setting('excludeaudiolang3') == "-----":
			langs.append(utils.langdict[setting('excludeaudiolang3')])
		availableaudio = xbmc.Player().getAvailableAudioStreams()
		debug('Available audio streams: %s' % availableaudio)
		availableaudio = " ".join(availableaudio)
		if any(x in availableaudio for x in langs):
			debug('Excluded: the audio language is excluded')
			return False
		if "und" in availableaudio and not boolsetting('audiound'):
			debug('Excluded: undertermined audio')
			return False
		return True
	return True


def subexclusion():

	if boolsetting('excludesub'):
		langs = []
		langs.append(utils.langdict[setting('excludesublang1')])
		if not setting('excludesublang2') == "-----":
			langs.append(utils.langdict[setting('excludesublang2')])
		if not setting('excludesublang3') == "-----":
			langs.append(utils.langdict[setting('excludesublang3')])
		availablesubs = xbmc.Player().getAvailableSubtitleStreams()
		debug('Available sub languages: %s' % availablesubs)
		availablesubs = " ".join(availablesubs)
		if any(x in availablesubs for x in langs):
			debug('Subtitle is already present')
			return False
		return True
	return True


def pathexclusion():

	debug('Content path: %s' % videopath())
	if videopath().find("http://") > -1 or videopath().find("https://") > -1:
		debug('Content source: %s' % videosource())

	if not videopath():
		return False

	if (videopath().find("pvr://") > -1) and boolsetting('excludelivetv'):
		debug('Video is playing via Live TV, which is currently set as excluded location.')
		return False

	if (videopath().find("http://") > -1 or videopath().find("https://") > -1) and boolsetting('excludehttp'):
		debug('Video is playing via HTTP or HTTPS source, which is currently set as excluded location.')
		return False

	path = setting('path')
	if path and boolsetting('excludepath'):
		if (videopath().find(path) > -1):
			debug('Video is playing from %s, which is currently set as excluded path 1.' % path)
			return False

	path2 = setting('path2')
	if path2 and boolsetting('excludepath2'):
		if (videopath().find(path2) > -1):
			debug('Video is playing from %s, which is currently set as excluded path 2.' % path2)
			return False

	path3 = setting('path3')
	if path3 and boolsetting('excludepath3'):
		if (videopath().find(path3) > -1):
			debug('Video is playing from %s, which is currently set as excluded path 3.' % path3)
			return False

	path4 = setting('path4')
	if path4 and boolsetting('excludepath4'):
		if (videopath().find(path4) > -1):
			debug('Video is playing from %s, which is currently set as excluded path 4.' % path4)
			return False

	path5 = setting('path5')
	if path5 and boolsetting('excludepath5'):
		if (videopath().find(path5) > -1):
			debug('Video is playing from %s, which is currently set as excluded path 5.' % path5)
			return False

	return True


def globalexclusion():

	if pathexclusion() and addonexclusion() and wordsexclusion() and timeexclusion() and subexclusion() and audioexclusion():
		return True
	return False

#********************************************************************************************************************************************************

resources/lib/opensubtitles.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

import os, sys, xbmc, xbmcaddon, xbmcgui
import base64, gzip, xmlrpclib, xbmcvfs
import StringIO, re, hashlib, json, urlparse
import codecs, urllib, xbmcplugin




def loadsub():

	try:
		addon = "service.subloader"#alterar quando estiver tudo ligado
		setting = xbmcaddon.Addon(addon).getSetting
		server = xmlrpclib.Server('http://api.opensubtitles.org/xml-rpc', verbose=0)
#		token = server.LogIn('', '', 'en', 'XBMC_Subtitles_v1')['token'] (token antigo)
		token = server.LogIn('', '', 'en', 'XBMC_Subtitles_Unofficial_v5.2.15')['token']
		media = xbmc.Player().getVideoInfoTag().getMediaType()




		langDict = {
			'Afrikaans': 'afr', 'Albanian': 'alb', 'Arabic': 'ara', 'Armenian': 'arm', 'Basque': 'baq',
			'Bengali': 'ben', 'Bosnian': 'bos', 'Breton': 'bre', 'Bulgarian': 'bul', 'Burmese': 'bur',
			'Catalan': 'cat', 'Chinese': 'chi', 'Croatian': 'hrv', 'Czech': 'cze', 'Danish': 'dan', 'Dutch': 'dut',
			'English': 'eng', 'Esperanto': 'epo', 'Estonian': 'est', 'Finnish': 'fin', 'French': 'fre',
			'Galician': 'glg', 'Georgian': 'geo', 'German': 'ger', 'Greek': 'ell', 'Hebrew': 'heb', 'Hindi': 'hin',
			'Hungarian': 'hun', 'Icelandic': 'ice', 'Indonesian': 'ind', 'Italian': 'ita', 'Japanese': 'jpn',
			'Kazakh': 'kaz', 'Khmer': 'khm', 'Korean': 'kor', 'Latvian': 'lav', 'Lithuanian': 'lit',
			'Luxembourgish': 'ltz', 'Macedonian': 'mac', 'Malay': 'may', 'Malayalam': 'mal', 'Manipuri': 'mni',
			'Mongolian': 'mon', 'Montenegrin': 'mne', 'Norwegian': 'nor', 'Occitan': 'oci', 'Persian': 'per',
			'Polish': 'pol', 'Portuguese': 'por', 'Portuguese(Brazil)': 'pob', 'Romanian': 'rum',
			'Russian': 'rus', 'Serbian': 'scc', 'Sinhalese': 'sin', 'Slovak': 'slo', 'Slovenian': 'slv',
			'Spanish': 'spa', 'Swahili': 'swa', 'Swedish': 'swe', 'Syriac': 'syr', 'Tagalog': 'tgl', 'Tamil': 'tam',
			'Telugu': 'tel', 'Thai': 'tha', 'Turkish': 'tur', 'Ukrainian': 'ukr', 'Urdu': 'urd'}

		codePageDict = {
			'ara': 'cp1256', 'ar': 'cp1256', 'ell': 'cp1253', 'el': 'cp1253', 'heb': 'cp1255',
			'he': 'cp1255', 'tur': 'cp1254', 'tr': 'cp1254', 'rus': 'cp1251', 'ru': 'cp1251'}

		release = ['cam', 'camrip', 'telesync', 'ts', 'hdts', 'hc', 'hcrip', 'pdvd', 'predvdrip', 
			'telecine', 'tc', 'hdtc', 'ppv', 'ppvrip', 'screener', 'scr', 'dvdscr', 'dvdcreener', 'bdscr', 
			'line', 'dvd', 'dvdrip', 'dvdr', 'dvdmux', 'dvdfull', 'hdtv', 'dsr', 'dsrip', 'satrip', 'dthrip', 
			'dvbrip', 'pdtv', 'dtvrip', 'tvrip', 'hdtvrip', 'vod', 'vodrip', 'vodr', 'webdl', 'webdlrip', 'web', 
			'hdrip', 'webrip', 'webr', 'webcap', 'bluray', 'bdrip', 'brip', 'brrip', 'bdmv', 'divx', 'xvid']

		secrelease = ['rip', 'dl', 'cap', 'blu', 'ray']

		langs = []

		filter = []
		


		if setting('subtitles') == 'true':
			langs.append(langDict[setting('subtitles.lang.1')])
			if not setting('subtitles.lang.2') == 'None':
				langs.append(langDict[setting('subtitles.lang.2')])
			if not setting('subtitles.lang.3') == 'None':
				langs.append(langDict[setting('subtitles.lang.3')])
		else:
			raise Exception()
			
		sublanguageid = ','.join(langs)




# Get IMDBNumber or Name ************************************************************************************************************

		imdb = xbmc.Player().getVideoInfoTag().getIMDBNumber()

		if imdb.ljust(2)[:2].strip() == 'tt':
			imdbid = re.sub('[^0-9]', '', imdb)
		else:
			imdbid = 'None'
			if media == 'movie' or media == 'video':
				title = (xbmc.Player().getVideoInfoTag().getOriginalTitle()).lower()
				year = xbmc.Player().getVideoInfoTag().getYear()
				query = title + " " + str(year)
			if media == 'episode':
				query = (xbmc.Player().getVideoInfoTag().getTVShowTitle()).lower()
			else:
				raise Exception()
		if media == 'episode':
			season = xbmc.Player().getVideoInfoTag().getSeason()
			episode = xbmc.Player().getVideoInfoTag().getEpisode()

#************************************************************************************************************************************



# Get release type ******************************************************************************************************************

		vidPath = xbmc.Player().getPlayingFile()
		fmt = re.split('[.:;()[\]{}\\\\/\s\&€\#\=\$\?\!%\+\-_\*[0-9]', vidPath)
		fmt = [i.lower() for i in fmt]
		fmt = [i for i in fmt if i in release]
		fmt = list(dict.fromkeys(fmt))
		fmtst = ''.join(fmt)
		secvidPath = vidPath.split(fmtst)[-1]
		secfmt = re.split('[.:;()[\]{}\\\\/\s\&€\#\=\$\?\!%\+\-_\*[0-9]', secvidPath)
		secfmt = [i.lower() for i in secfmt]
		secfmt = [i for i in secfmt if i in secrelease]
		secfmt = list(dict.fromkeys(secfmt))
		fmt = fmt + secfmt

		if fmt:
			fmtst = fmtst + (''.join(secfmt))
		else:
			if setting('notif') == 'true' and setting('subsearch') == 'true':
				xbmc.executebuiltin('Notification("%s", "%s", "%s",)' % ("Opening sub dialog", "No release type detected", 5000))
				raise Exception()
			else:
				if setting('notif') == 'true':
					xbmc.executebuiltin('Notification("%s", "%s", "%s",)' % ("Aborting", "No release type detected", 5000))
					raise Exception()
				else:
					raise Exception()

#************************************************************************************************************************************




		if fmtst == 'cam' or fmtst == 'camrip':
			fmt = ['cam-rip', 'cam.rip', 'cam rip', 'cam', 'camrip']
			fmtflex = ['cam-rip', 'cam.rip', 'cam rip', 'cam', 'camrip', 'telesync', 'ts', 'hdts', 'hc', 'hcrip', 'pdvd', 'predvdrip']
		if fmtst == 'telesync' or fmtst == 'ts' or fmtst == 'hdts' or fmtst == 'hc' or fmtst == 'hcrip' or fmtst == 'pdvd' or fmtst == 'predvdrip':
			fmt = ['telesync', 'ts', 'hdts', 'hc', 'hcrip', 'pdvd', 'predvdrip']
			fmtflex = ['cam-rip', 'cam.rip', 'cam rip', 'cam', 'camrip', 'telesync', 'ts', 'hdts', 'hc', 'hcrip', 'pdvd', 'predvdrip']
		if fmtst == 'telecine' or fmtst == 'tc' or fmtst == 'hdtc':
			fmt = ['telecine', 'tc', 'hdtc']
			fmtflex = fmt
		if fmtst == 'ppv' or fmtst == 'ppvrip':
			fmt = ['ppv', 'ppvrip']
			fmtflex = fmt
		if fmtst == 'screener' or fmtst == 'scr' or fmtst == 'dvdscr' or fmtst == 'dvdcreener' or fmtst == 'bdscr':
			fmt = ['screener', 'scr', 'dvdscr', 'dvdcreener', 'bdscr']
			fmtflex = ['screener', 'scr', 'dvdscr', 'dvdcreener', 'bdscr', 'dvd', 'dvd-r', 'dvdrip', 'dvd rip', 'dvdr', 'dvdmux', 'dvdfull', 'dvd5', 'dvd9', 'dvd-rip', 'dvd.rip', 'dvd-mux', 'dvd.mux', 'dvd-5', 'dvd-9']
		if fmtst == 'line':
			fmt = ['r5', 'r5line', 'line', 'r5.line', 'r5-line', 'r5 line']
			fmtflex = fmt
		if fmtst == 'dvd' or fmtst == 'dvdrip' or fmtst == 'dvdr' or fmtst == 'dvdmux' or fmtst == 'dvdfull':
			fmt = ['dvd', 'dvd-r', 'dvdrip', 'dvdr', 'dvdmux', 'dvdfull', 'dvd5', 'dvd9', 'dvd-rip', 'dvd rip', 'dvd.rip', 'dvd-mux', 'dvd.mux', 'dvd-5', 'dvd-9']
			fmtflex = ['screener', 'scr', 'dvdscr', 'dvdcreener', 'bdscr', 'dvd', 'dvd-r', 'dvdrip', 'dvd rip', 'dvdr', 'dvdmux', 'dvdfull', 'dvd5', 'dvd9', 'dvd-rip', 'dvd.rip', 'dvd-mux', 'dvd.mux', 'dvd-5', 'dvd-9']
		if fmtst == 'hdtv' or fmtst == 'dsr' or fmtst == 'dsrip' or fmtst == 'satrip' or fmtst == 'dthrip' or fmtst == 'dvbrip' or fmtst == 'pdtv' or fmtst == 'dtvrip' or fmtst == 'tvrip' or fmtst == 'hdtvrip':
			fmt = ['hdtv', 'dsr', 'dsrip', 'satrip', 'dthrip', 'dvbrip', 'pdtv', 'dtvrip', 'tvrip', 'hdtvrip', 'hdtv-rip', 'hdtv.rip', 'hdtv rip']
			fmtflex = fmt
		if fmtst == 'vob' or fmtst == 'vodrip' or fmtst == 'vodr':
			fmt = ['vob', 'vodrip', 'vod rip', 'vod.rip', 'vod-rip', 'vodr']
			fmtflex = fmt
		if fmtst == 'webdl' or fmtst == 'webdlrip' or fmtst == 'dl':
			fmt = ['web-dl', 'webdl', 'web.dl', 'web dl', 'webdlrip']
			fmtflex = ['web-dl', 'webdl', 'web.dl', 'web dl', 'webdlrip', 'webrip', 'web-rip', 'web.rip', 'web rip', 'webr', 'web', 'webcap', 'web-cap', 'web.cap', 'web cap', 'hdrip']
		if fmtst == 'webrip' or fmtst == 'webr' or fmtst == 'webcap' or fmtst == 'web' or fmtst == 'cap':
			fmt = ['webrip', 'web-rip', 'web.rip', 'web rip', 'webr', 'web', 'webcap', 'web-cap', 'web.cap', 'web cap']
			fmtflex = ['web-dl', 'webdl', 'web.dl', 'web dl', 'webdlrip', 'webrip', 'web-rip', 'web.rip', 'web rip', 'webr', 'web', 'webcap', 'web-cap', 'web.cap', 'web cap', 'hdrip']
		if fmtst == 'bluray' or fmtst == 'blurayrip' or fmtst == 'blu' or fmtst == 'ray':
			fmt = ['blu', 'ray', 'bluray', 'blu-ray', 'blu.ray', 'blu ray']
			fmtflex = ['blu', 'ray', 'bluray', 'blu-ray', 'blu.ray', 'blu ray', 'bdrip', 'bd-rip', 'bd.rip', 'bd rip', 'brrip', 'br-rip', 'br.rip', 'br rip', 'brip', 'b-rip', 'b.rip', 'b rip', 'bdmv']
		if fmtst == 'bdrip' or fmtst == 'brip' or fmtst == 'brrip' or fmtst == 'bdmv':
			fmt = ['bdrip', 'bd-rip', 'bd.rip', 'bd rip', 'brrip', 'br-rip', 'br.rip', 'br rip', 'brip', 'b-rip', 'b.rip', 'b rip', 'bdmv']
			fmtflex = ['blu', 'ray', 'bluray', 'blu-ray', 'blu.ray', 'blu ray', 'bdrip', 'bd-rip', 'bd.rip', 'bd rip', 'brrip', 'br-rip', 'br.rip', 'br rip', 'brip', 'b-rip', 'b.rip', 'b rip', 'bdmv']
		if fmtst == 'hdrip':
			fmt = ['hdrip', 'hd-rip', 'hd rip', 'hd.rip']
			fmtflex = ['web-dl', 'webdl', 'web.dl', 'web dl', 'webdlrip', 'hdrip', 'webrip', 'web-rip', 'web.rip', 'web rip', 'webr', 'web', 'webcap', 'web-cap', 'web.cap', 'web cap']








		filelocal = xbmc.getInfoLabel('Player.Filenameandpath')#funciona local
		filestr = vidPath.split('/')[-1]#funciona em cache torrents seren





		if setting('notif') == 'true':
			xbmc.executebuiltin('Notification("%s", "%s", "%s",)' % (fmtst, filestr, 4000))




		if media == 'episode':
			if imdbid is not 'None':
				result = server.SearchSubtitles(token, [{'sublanguageid': sublanguageid, 'imdbid': imdbid, 'season': season, 'episode': episode}])['data']
			else:
				result = server.SearchSubtitles(token, [{'sublanguageid': sublanguageid, 'query': query, 'season': season, 'episode': episode}])['data']
		else:
			if imdbid is not 'None':
				result = server.SearchSubtitles(token, [{'sublanguageid': sublanguageid, 'imdbid': imdbid}])['data']
			else:
				result = server.SearchSubtitles(token, [{'sublanguageid': sublanguageid, 'query': query}])['data']






		result = [i for i in result if i['SubSumCD'] == '1']
	
	
		for lang in langs:
			filter += [i for i in result if i['SubLanguageID'] == lang and any(x in i['MovieReleaseName'].lower() for x in fmt)]
#			filter += [i for i in result if i['SubLanguageID'] == lang and any(x in i['MovieReleaseName'].lower() for x in quality)]
#			filter += [i for i in result if i['SubLanguageID'] == lang]


		try:
			lang = filter[0]['SubLanguageID']
		except Exception:
			if setting('strict') == 'true':
				raise Exception()
			else:
				for lang in langs:
					filter += [i for i in result if i['SubLanguageID'] == lang and any(x in i['MovieReleaseName'].lower() for x in fmtflex)]
					lang = filter[0]['SubLanguageID']
# colocar aqui um raise exception e criar log...




#		try:
#			lang = xbmc.convertLanguage(filter[0]['SubLanguageID'], xbmc.ISO_639_1)
#		except Exception:
#			lang = filter[0]['SubLanguageID']




#		lang = filter[0]['SubLanguageID']
		content = [filter[0]['IDSubtitleFile'], ]
		content = server.DownloadSubtitles(token, content)
		content = base64.b64decode(content['data'][0]['data'])
		content = gzip.GzipFile(fileobj=StringIO.StringIO(content)).read()

		subtitle = xbmc.translatePath('special://temp/')
		subtitle = os.path.join(subtitle, 'TemporarySubs.%s.srt' % lang)

		file = xbmcvfs.File(subtitle, 'w')
		file.write(str(content))
		file.close()

		xbmc.sleep(1000)
		xbmc.Player().setSubtitles(subtitle)
	

		if setting('notif') == 'true':
			xbmc.sleep(4000)
			test = [filter[0]['MovieReleaseName'], ]
			xbmc.executebuiltin('Notification("%s", "%s", "%s",)' % (lang, test, 4000))

	except Exception:
		if setting('subsearch') == 'true':
			xbmc.executebuiltin('XBMC.ActivateWindow(SubtitleSearch)')
